lib/std/debug.zig

```diff
@@ -224,24 +529,49 @@ pub fn assert(ok: bool) void {
 }
 
 pub fn panic(comptime format: []const u8, args: anytype) noreturn {
+    panicExtra(null, format, args);
+}
+
+/// `panicExtra` is useful when you want to print out an `@errorReturnTrace`
+/// and also print out some values.
+pub fn panicExtra(
+    trace: ?*builtin.StackTrace,
+    comptime format: []const u8,
+    args: anytype,
+) noreturn {
     @setCold(true);
-    // TODO: remove conditional once wasi / LLVM defines __builtin_return_address
-    const first_trace_addr = if (builtin.os.tag == .wasi) null else @returnAddress();
-    panicExtra(null, first_trace_addr, format, args);
+    const size = 0x1000;
+    const trunc_msg = "(msg truncated)";
+    var buf: [size + trunc_msg.len]u8 = undefined;
+    // a minor annoyance with this is that it will result in the NoSpaceLeft
+    // error being part of the @panic stack trace (but that error should
+    // only happen rarely)
+    const msg = std.fmt.bufPrint(buf[0..size], format, args) catch |err| switch (err) {
+        std.fmt.BufPrintError.NoSpaceLeft => blk: {
+            std.mem.copy(u8, buf[size..], trunc_msg);
+            break :blk &buf;
+        },
+    };
+    builtin.panic(msg, trace);
 }
 
 /// Non-zero whenever the program triggered a panic.
-/// The counter is incremented/decremented atomically.
-var panicking: u8 = 0;
-
-// Locked to avoid interleaving panic messages from multiple threads.
+/// Counts the number of threads which are waiting on the `panic_mutex`
+/// to print out the stack traces and msg.
+var panicking: std.atomic.Int(u32) = std.atomic.Int(u32).init(0);
```

Comment:
This used to be a `u8`, but it was changed to a `u32` based on some discussions in discord. I don't believe this reduces the number of platforms on which `panicImpl` is supported (I think everything which supports atomic `u8` operations also supports atomic `u32` operations.

Reply:
TLDR on discord discussion: u8 could overflow due to the mutex which would result in panic messages being cutoff, also there isn't much cost to u32. I ended up going with u32 over usize because I though this reflected the intention somewhat better.
Full discord discussion:

> Ryan Greenblatt — Today at 1:14 PM
> Yep. I will use a u8.
> 
> Protty — Today at 1:31 PM
> That could potentially overflow if theres a lot of threads
> 
> Ryan Greenblatt — Today at 1:49 PM
> Well, its a u8 now. I could make it a u32. I don't see any real downside to doing that.
> The edge case in which that value overflows seems very extreme to me.
> You need so many simultaneous panics with exactly the right timing.
> Also, it won't panic on overflow because the atomic intrinsics have well defined overflow behavior (wrap)
> So, the only issue is that you might have a thread abort while another thread is printing out the panic message.
> 
> fengb — Today at 1:58 PM
> What if we have 256 threads panicking at once :thinking:
> 
> Protty — Today at 2:13 PM
> `@Ryan Greenblatt` the silent unpanicing overflow is the issue itself as it would mess up the synchronization of calling abort() and potentially abort() too early. 
> 
> The case of this is not as unrealistic as it sounds givrn the criticial section for the threads involves acquiring a mutex. If a few hundred threads panic at the same time and concurrently hit that codepath, it wouldnt finish printing out all their stack traces
> 
> Ryan Greenblatt — Today at 2:18 PM
> Yeah, I switched it to a u32. I think that should be big enough...
> 
> Protty — Today at 2:19 PM
> Practically equivalent, although id do usize in general
> 
> Ryan Greenblatt — Today at 2:20 PM
> is std.atomic.Int(usize) guaranteed to exist?
> Or std.atomic.Int(u32) for that matter.
> 
> Protty — Today at 2:20 PM
> Dont think anything isnreally guanrateed to exist atm but using std stuff is generally recommended over the builtin counterparts. Think thats what most builtin stuff are moving to
> 
> Ryan Greenblatt — Today at 2:21 PM
> Yeah, I mean are there platforms which don't support atomic operations on the size type?
> 
> Protty — Today at 2:22 PM
> Ik gba and some resitrcted mips dont. But those platforms wouldnt be using the default panic handler
> A lot of std default impls wouldnt work for them which is fine